docs/architecture/adr-067-mempool-refactor.md

```diff
@@ -0,0 +1,215 @@
+# ADR 067: Mempool Refactor
+
+- [ADR 067: Mempool Refactor](#adr-067-mempool-refactor)
+  - [Changelog](#changelog)
+  - [Status](#status)
+  - [Context](#context)
+  - [Alternative Approaches](#alternative-approaches)
+  - [Decision](#decision)
+  - [Detailed Design](#detailed-design)
+    - [CheckTx](#checktx)
+    - [Mempool](#mempool)
+    - [Eviction](#eviction)
+    - [Gossiping](#gossiping)
+  - [Consequences](#consequences)
+    - [Positive](#positive)
+    - [Negative](#negative)
+    - [Neutral](#neutral)
+  - [References](#references)
+
+## Changelog
+
+- April 19, 2021: Initial Draft (@alexanderbez)
+
+## Status
+
+Proposed
+
+## Context
+
+Tendermint Core has a reactor and data structure, mempool, that facilitates the
+ephemeral storage of uncommitted transactions. Honest nodes participating in a
+Tendermint network gossip these uncommitted transactions to each other if they
+pass the application's `CheckTx`. In addition, block proposers select from the
+mempool a subset of uncommitted transactions to include in the next block.
+
+Currently, the mempool in Tendermint Core is designed as a FIFO queue. In other
+words, transactions are included in blocks as they are received by a node. There
+currently is no explicit and prioritized ordering of these uncommitted transactions.
+This presents a few technical and UX challenges for operators and applications.
+
+Namely, validators are not able to prioritize transactions by their fees or any
+incentive aligned mechanism. In addition, the lack of prioritization also leads
+to cascading effects in terms of DoS and various attack vectors on networks,
+e.g. [cosmos/cosmos-sdk#8224](https://github.com/cosmos/cosmos-sdk/discussions/8224).
+
+Thus, Tendermint Core needs the ability for an application and its users to
+prioritize transactions in a flexible and performant manner. Specifically, we're
+aiming to either improve, maintain or add the following properties in the
+Tendermint mempool:
+
+- Allow application-determined transaction priority.
+- Allow efficient concurrent reads and writes.
+- Allow block proposers to reap transactions efficiently by priority.
```

Comment:
Another key property here is that when block proposers reap transactions by priority, the result needs to be a valid block (ie. a block where all the txs pay fees).

This is preserved by the mempool today, but if we execute CheckTxs in one order, and put them in a block in a different order (priority), we still need to ensure that sequence is valid. 

We could do this by re-running CheckTx in the new order (which may not be that costly if all the accounts are cached and the signature checks are memoized), but would we then just drop any txs that don't pass? PrepareProposal of ABCI++ also gets the application to do the work, but in the meantime I think with the current proposal with the current SDK we could end up with correct proposers proposing invalid blocks since there could be dependent txs in the mempool.   

If we don't allow any dependent txs in the mempool (ie. CheckTx always runs against the latest committed state), then a single sender can spam the mempool with tons of mutually exclusive txs. So at least we'd need to track dependent txs per sender to prevent them spamming

Reply:
Ok for some reason I thought the SDK currently supported dependent txs between accounts that started with 0 balance like A->B->C->D but that's not the case since no Msgs run on CheckTx, so balances in the CheckTx state in the app should only be able to decrease. So long as that's true (I guess this is an invariant that would have to be stated for all ABCI apps to use the new mempool), and we correctly order txs per sender, then txs that pass CheckTx should still be valid in the block even if re ordered by priority.

In any case, we should include this order-independent validity criteria as a Property

Reply:
And I guess another thing to confirm is that this design will work without any major changes to the SDK - it does seem like it (other than populating the new fields in ResponseCheckTx)